src/server.mjs

```diff
@@ -1094,17 +1118,148 @@ api.post("/getPathNames", (req, res, next) => {
   });
 });
 
-api.post("/getBedRegions", (req, res) => {
+// returns whether or not a string is a valid http url
+function isValidURL(string) {
+  if (!string) {
+    return False;
+  }
+
+  let url;
+
+  try {
+    url = new URL(string)
+  } catch(error) {
+    return false;
+  }
+
+  return url.protocol === "http:" || url.protocol === "https:";
+}
+
+// given the datapath and url, download file to the datapath
+const downloadFile = async(fileURL, destinationPath, fileName) => {
+
+  if (!fileName) {
+    return;
+  }
+
+  const response = await fetchAndValidate(fileURL, config.maxFileSizeBytes);
+  
+  const destinationFile = path.resolve(destinationPath, fileName);
+
+  // overwrites file if it already exists
+  const fileStream = fs.createWriteStream(destinationFile, { flags: 'w' });
+  await finished(Readable.fromWeb(response.body).pipe(fileStream));
+
+}
+
+const fetchAndValidate = async(url, maxBytes) => {
+  const options = {
+    method: "GET",
+    credentials: "omit",
+    cache: "default",
+    signal: timeoutController(config.fetchTimeout).signal
+  };
+
+  let response = await fetch(url, options);
+
+  // check for unsuccessful response codes
+  if (!response.ok) {
+    throw new BadRequestError("Fetch request failed: " + response.status);
+  }
+
+  // check for size specified in header
+  const contentLength = response.headers.get("Content-Length");
+
+  if (contentLength > maxBytes) {
+    throw new Error("Fetch request failed: contentLength exceeds maximum file size");
+  }
+
+  // use a reader to make sure we're not reading past the max size allowed
+  const reader = response.body.getReader();
+
+  let bytesRead = 0;
+  const dataRead = [];
+
+  while (true) {
+    let {done, value} = await reader.read();
+
+    if (done) {
+      break;
+    }
+
+    dataRead.push(value);
+    bytesRead += new Blob([value]).size;
+
+    if (bytesRead > maxBytes) {
+      reader.cancel();
+      throw new Error("Fetched request failed, maximum file size exceeded: " + maxSize);
+    }
+  }
+
+  return new Response(new Blob(dataRead), { headers: response.headers });
```

Comment:
Maybe the interface of this function should change to just return a Blob?

Alternately, if we want to support downloading files that are larger than is convenient to hold in memory, but still use a size limit, we'd want to return a stream from this function, but with a size-limiting filter on the stream that will raise an error when the reader goes to read past the size limit.